src/H5VLnative_dataset.c

```diff
@@ -755,53 +756,75 @@ H5VL__native_dataset_optional(void *obj, H5VL_optional_args_t *args, hid_t dxpl_
         /* H5Dread_chunk */
         case H5VL_NATIVE_DATASET_CHUNK_READ: {
             H5VL_native_dataset_chunk_read_t *chunk_read_args = &opt_args->chunk_read;
-            hsize_t offset_copy[H5O_LAYOUT_NDIMS]; /* Internal copy of chunk offset */
 
             /* Check arguments */
             if (NULL == dset->oloc.file)
                 HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "dataset is not associated with a file");
 
-            /* Make sure the dataset is chunked or sparse chunk */
-            /* TBD: H5Dread_chunk() is supposed to work for sparse chunk */
-            if (H5D_CHUNKED != dset->shared->layout.type)
-                HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a chunked dataset");
-
-            /* Copy the user's offset array so we can be sure it's terminated properly.
-             * (we don't want to mess with the user's buffer).
-             */
-            if (H5D__chunk_get_offset_copy(dset, chunk_read_args->offset, offset_copy) < 0)
-                HGOTO_ERROR(H5E_DATASET, H5E_CANTCOPY, FAIL, "failure to copy offset array");
+            /* Check for shared chunk cache support */
+            if (dset->shared->layout.sc_ops) {
+                H5D_chunk_scc_udata_t udata;
+                size_t                buf_size = SIZE_MAX;
+                assert(0 && "set up the buf size from API when available. Either point directly into args "
+                            "struct in call to H5SC_direct_chunk_read() or fill in args struct afterwards");
+
+                /* Set up udata */
+                udata.filters = 0;
+                udata.size    = 0;
+
+                /* Dispatch call to H5SC layer */
+                if (H5SC_direct_chunk_read(H5F_get_shared_cache(dset->oloc.file), dset,
+                                           chunk_read_args->offset, &udata, chunk_read_args->buf,
+                                           &buf_size) < 0)
+                    HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL, "can't read unprocessed chunk data");
+
+                /* Return info to caller */
+                chunk_read_args->filters = udata.filters;
+            }
+            else {
+                /* Make sure the dataset is chunked */
+                if (H5D_CHUNKED != dset->shared->layout.type)
+                    HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a chunked dataset");
 
-            /* Read the raw chunk */
-            if (H5D__chunk_direct_read(dset, offset_copy, &chunk_read_args->filters, chunk_read_args->buf) <
-                0)
-                HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL, "can't read unprocessed chunk data");
+                /* Read the raw chunk */
+                if (H5D__chunk_direct_read(dset, chunk_read_args->offset, &chunk_read_args->filters,
```

Comment:
Checking the validity of the offset is layout specific and has been moved to H5Dchunk.c for the chunked non-SCC case.